src/components/app/ServiceWorkerManager.js

```diff
@@ -66,75 +130,198 @@ class ServiceWorkerManager extends PureComponent<Props, State> {
+        // is ready -- which means we won't need to download anything more.
+        return this._isProfileLoadedAndReady();
       default:
         throw assertExhaustiveCheck(dataSource);
     }
+  }
 
-    if (installStatus !== 'ready') {
-      return null;
+  componentDidMount() {
+    if (process.env.NODE_ENV === 'production') {
+      this._installServiceWorker();
+    }
+  }
+
+  componentDidUpdate() {
+    const { phase } = this.props;
+    const { installStatus } = this.state;
+
+    if (installStatus !== 'idle' && phase === 'FATAL_ERROR') {
+      // If we got a fatal error and a new version of the application is
+      // available, let's try to reload automatically, as this might fix the
+      // fatal error.
+      const runtime = require('offline-plugin/runtime');
+      runtime.applyUpdate();
+      this.reloadPage();
     }
+  }
+
+  applyServiceWorkerUpdate = () => {
+    this.setState({ installStatus: 'installing' });
+    const runtime = require('offline-plugin/runtime');
+    runtime.applyUpdate();
+  };
+
+  reloadPage = () => {
+    window.location.reload();
+  };
+
+  _onCloseNotice = () => {
+    this.setState({ isNoticeDisplayed: false });
+  };
+
+  renderButton() {
+    const { installStatus } = this.state;
+
+    switch (installStatus) {
+      case 'installing':
+        return (
+          <button
+            className="photon-button photon-button-micro photon-message-bar-action-button"
+            type="button"
+          >
+            Installing…
+          </button>
+        );
+      case 'idle':
+        throw new Error(
+          `We tried to render the notice's button but because we're in status 'idle', there should be no notice.`
+        );
+      case 'pending':
+        return (
+          <button
+            className="photon-button photon-button-micro photon-message-bar-action-button"
+            type="button"
+            onClick={this.applyServiceWorkerUpdate}
+          >
+            Apply and reload
+          </button>
+        );
+      case 'installed':
+        // Another tab applied the new service worker.
+        return (
+          <button
+            className="photon-button photon-button-micro photon-message-bar-action-button"
+            type="button"
+            onClick={this.reloadPage}
+          >
+            Reload the application
+          </button>
+        );
+      default:
+        throw assertExhaustiveCheck(installStatus);
+    }
+  }
+
+  render() {
+    const { isNoticeDisplayed, updatedWhileNotReady } = this.state;
 
     if (!isNoticeDisplayed) {
       return null;
     }
 
+    if (updatedWhileNotReady) {
+      // The user updated the service worker from another tab, before this tab
+      // was fully loaded. There may be problems if we need to load some
+      // resources, for example the demandling wasm module.
+      return (
+        <div className="serviceworker-ready-notice-wrapper">
```

Comment:
This would be nice to try out the code, but it seems hard to trigger to review.

Reply:
Yeah, I don't know how to trigger this from the deploy preview. I think you'd have to checkout locally to tweak the states.